### PR TITLE
[CPU] Fix memory access for sbgemm bf16

### LINKED ISSUE #156022: [Segfault Bug] Out-of-bounds write of at::native::cpubla::gemm (bfloat16) in at::native::cpu_flash_attention_backward
### 🐛 Describe the bug

In PyTorch’s CPU implementation of [cpu_flash_attention_backward](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cpu/FlashAttentionKernel.cpp#L745), the [cpublas::gemm](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/CPUBlas.cpp#L329) performs out-of-bounds memory writes when it's dispatched to **openblas sbgemm_** routine. 

This results causes **nondeterministic crashes** and **memory corruption**. The fault is observed in: `aten/src/ATen/native/cpu/FlashAttentionKernel.cpp::cpu_flash_attention_backward`. 

this bug can be reproduced using pytorch unit tests of `TestSDPACpuOnlyCPU.test_scaled_dot_product_fused_attention`

```bash
for i in {1..1000}
do
python test/test_transformers.py TestSDPACpuOnlyCPU.test_scaled_dot_product_fused_attention_mask_vs_math_cpu_fused_kernel0_bfloat16_batch_size_12_q_seq_len_1030_kv_seq_len_17_n_head_3_head_dim_8_mask_dim_2_bool_mask_False_train_True_casual_False_set_attn_mask_False_cpu_bfloat16
done
```

#### Root cause and Correction of cpublas::gemm

This segfault bug is caused by incorrect behavior of [bf16 cpublas::gemm](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/CPUBlas.cpp#L356-L373), which mishandles strided output matrices.

Specifically, it:

1. Allocates a temporary FP32 buffer of size ldc * n
2. Copies and converts the full ldc * n region from the output matrix C (BF16 → FP32)
3. Performs sbgemm over the logical m × n region using the passed ldc
4. Converts the entire ldc * n result back to BF16 and writes it back to C

This logic assumes C is contiguous or has exclusive access to ldc * n elements, which is not valid when C is a view into a larger tensor (as in multi-head attention). As a result, it overwrites memory outside the m × n region, corrupting adjacent data.

**Expected behavior:**
cpublas::gemm should only copy, compute, and write back the m × n region as described by the passed dimensions and ldc stride, not the full ldc × n span.

### Error Explanation of flash attention:

The bug occurs in [`cpu_flash_attention_backward`](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cpu/FlashAttentionKernel.cpp#L745), where `cpublas::gemm` is called on `grad_q`, `grad_k`, and `grad_v` which are of strided views.

Take `grad_q` as an example. Its shape is: `[batchSize, qSize, numHeads, headSize]`. Each attention head is processed as a logical matrix: `[qSize, headSize]`.  But in memory, **each row is spaced by `numHeads * headSize` elements**, not just `headSize`. That is: `ldc = numHeads * headSize > m = headSize`. 

This layout is safe for the first head (`head == 0`), where writes begin at the start of the grad tensor. However, for **`head > 0`**, the output pointer `C` starts at an offset in the middle of the tensor. Since `cpublas::gemm` writes back a full `ldc * n` region from that offset, it overruns the intended slice and **corrupts memory belonging to other heads or unrelated data**.

| Head Index | Behavior               |
|------------|------------------------|
| ✅ `head == 0` | Writes start at beginning — overflow is hidden |
| ❌ `head > 0`  | Writes past intended slice — causes memory corruption |

### Proposed fixup:

code in this https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/CPUBlas.cpp#L356-L373 should be changed to 

```
   if (use_blas_gemm(transa, transb, m, n, k, lda, ldb, ldc)) {
      int m_ = m, n_ = n, k_ = k, lda_ = lda, ldb_ = ldb, ldc_ = ldc;
      char transa_ = to_blas(transa), transb_ = to_blas(transb);
      float alpha_ = alpha, beta_ = beta;
      int c_size = n_ * m_;
      // C matrix in OpenBLAS sbgemm are of type "float" so we have to convert, copy and copy back.
      at::BFloat16 * c_ptr = c; 
      std::vector<float> float_v(c_size, 0.0f);
      for (const auto j : c10::irange(n)) {
        for (const auto i : c10::irange(m)) {
          float_v[j * m + i] = c10::convert<float>(*(c_ptr + i));
        }
        c_ptr += ldc_;
      }
      sbgemm_(&transa_, &transb_,
              &m_, &n_, &k_,
              &alpha_,
              a, &lda_,
              b, &ldb_,
              &beta_,
              float_v.data(), &m_);
      c_ptr = c;
      for (const auto j : c10::irange(n)) {
        for (const auto i : c10::irange(m)) {
          *(c_ptr + i) = c10::convert<at::BFloat16>(float_v[j * m + i]);
        }
        c_ptr += ldc_;
      }
      return;
   }
```

### Versions

```
Collecting environment information...
PyTorch version: 2.8.0a0+git655b3b1
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.1 LTS (aarch64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: version 4.0.2
Libc version: glibc-2.39

Python version: 3.12.3 (main, Feb  4 2025, 14:48:35) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-1029-aws-aarch64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         aarch64
CPU op-mode(s):                       64-bit
Byte Order:                           Little Endian
CPU(s):                               16
On-line CPU(s) list:                  0-15
Vendor ID:                            ARM
Model name:                           Neoverse-V2
Model:                                1
Thread(s) per core:                   1
Core(s) per socket:                   16
Socket(s):                            1
Stepping:                             r0p1
BogoMIPS:                             2000.00
Flags:                                fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 asimddp sha512 sve asimdfhm dit uscat ilrcpc flagm ssbs sb paca pacg dcpodp sve2 sveaes svepmull svebitperm svesha3 flagm2 frint svei8mm svebf16 i8mm bf16 dgh rng bti
L1d cache:                            1 MiB (16 instances)
L1i cache:                            1 MiB (16 instances)
L2 cache:                             32 MiB (16 instances)
L3 cache:                             36 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-15
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; __user pointer sanitization
Vulnerability Spectre v2:             Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] optree==0.15.0
[pip3] torch==2.8.0a0+git1bebe04
[conda] Could not collec

```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168

### PR DESCRIPTION
**Pull request by Valentine233**
Fixes #156022.

1. The original dtype conversion overwrites the whole `n_*ldc_` instead of `n_*m_` with stride `ldc_`, causing the potential memory issue.
2. Fix the None value issue in attention backward UT, as the sbgemm bf16 could be used.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168

### CODE REVIEW

**Comment by taoye9** [comment]
@Valentine233 hi, thanks for pushing the code. looks like there is a ci failure.

**Comment by taoye9** [comment]
@Ryo-not-rio i think you've already pushed code to another pr? Do you observe the same failure?

**Comment by Valentine233** [comment]
> @Valentine233 hi, thanks for pushing the code. looks like there is a ci failure.

That is an unrelated failure.
image

**Comment by Valentine233** [comment]
> e already pushed code to another pr? Do you observe the same failure?

Do you mean there is already a similar PR to fix the issue?

**Comment by taoye9** [comment]
> > e already pushed code to another pr? Do you observe the same failure?
> 
> Do you mean there is already a similar PR to fix the issue?

sry, my mistake. after double check, there isn't any other pr trying to fix this problem.

**Comment by taoye9** [comment]
hi, looks like the ci fails on [cuda12.8-py3.10-gcc9-sm75 / test (pr_time_benchmarks, 1, 1, linux.g4dn.metal.nvidia.gpu, unstable)](https://github.com/pytorch/pytorch/actions/runs/15818507209/job/44647824897?pr=156585#logs)

```
WIN: benchmark ('aotdispatcher_partitioner_cpu2', 'compile_time_instruction_count') failed, actual result 1928952840 is -1.73% lower than expected 1963000000 ±1.50% please OPEN THE TEST RESULTS update ALL BENCHMARKS RESULT with the new printed expected results. ALL ALL ALL

please update all results that changed significantly, and not only the failed ones
```

@Valentine233

**Comment by Valentine233** [comment]
> hi, looks like the ci fails on [cuda12.8-py3.10-gcc9-sm75 / test (pr_time_benchmarks, 1, 1, linux.g4dn.metal.nvidia.gpu, unstable)](https://github.com/pytorch/pytorch/actions/runs/15818507209/job/44647824897?pr=156585#logs)
> 
> ```
> WIN: benchmark ('aotdispatcher_partitioner_cpu2', 'compile_time_instruction_count') failed, actual result 1928952840 is -1.73% lower than expected 1963000000 ±1.50% please OPEN THE TEST RESULTS update ALL BENCHMARKS RESULT with the new printed expected results. ALL ALL ALL
> 
> please update all results that changed significantly, and not only the failed ones
> ```
> 
> @Valentine233

This is an unrelated failure. Other PRs also encounter this.

**Review by mingfeima — CHANGES_REQUESTED** [review]
update the test case. This is the slow fallback path, can we make sure it is tested in torch CI?

**Comment by taoye9** [comment]
hi @Valentine233 here is some update:

1. i run the sdpa ut test on local arm machine (neoversev1, and neoversev2), the segfault is gone.
2. i also managed to reproduced numeric failure. However, changing random see in torch.manual_seed(777) can solve the problem. so this might be a hint that numeric stability is of problem.

I wonder should we loose the check a bit or spent more time digging this problem? would like to hear your opinions @aditew01 @Valentine233 

```
Mismatched elements: 1 / 4896 (0.0%)
Greatest absolute difference: 0.0546875 at index (4, 0, 2, 1) (up to 0.01 allowed)
Greatest relative difference: 0.0333251953125 at index (4, 0, 2, 1) (up to 0.01 allowed)
```

**Comment by Valentine233** [comment]
> hi @Valentine233 here is some update:
> 
> 1. i run the sdpa ut test on local arm machine (neoversev1, and neoversev2), the segfault is gone.
> 2. i also managed to reproduced numeric failure. However, changing random see in torch.manual_seed(777) can solve the problem. so this might be a hint that numeric stability is of problem.
> 
> I wonder should we loose the check a bit or spent more time digging this problem? would like to hear your opinions @aditew01 @Valentine233
> 
> ```
> Mismatched elements: 1 / 4896 (0.0%)
> Greatest absolute difference: 0.0546875 at index (4, 0, 2, 1) (up to 0.01 allowed)
> Greatest relative difference: 0.0333251953125 at index (4, 0, 2, 1) (up to 0.01 allowed)
> ```

Thanks! I loose the tolerance value for backward, as 1) backward is supposed to have a larger numeric diff compared to forward, 2) the number of mismatched element is very few so this is a stability problem.

**Comment by aditew01** [comment]
Thanks for the PR @Valentine233 . The UT will only hit the code for aarch64 once this PR is merged: https://github.com/pytorch/pytorch/pull/151547 . (it enables the said sbgemm_ kernel which is used here)

**Comment by Valentine233** [comment]
> Thanks for the PR @Valentine233 . The UT will only hit the code for aarch64 once this PR is merged: #151547 . (it enables the said sbgemm_ kernel which is used here)

Seems https://github.com/pytorch/pytorch/pull/151547 also changes `aten/src/ATen/native/CPUBlas.cpp`, do you still need the related modification in this PR? The modification in this PR is more efficient.

For aarch64, why will it go into `sbgemm_` instead of `mkldnn_bf16_gemm`? Is `mkldnn` always disabled for aarch64?

**Comment by aditew01** [comment]
> Is mkldnn always disabled for aarch64?

That's a very good question, at first glance that was my assumption as well. 
However, in SDPA backward pass, [this](https://github.com/pytorch/pytorch/blob/6085120c01e1c0b51a6e27872b9b40a6cce0644f/aten/src/ATen/native/CPUBlas.cpp#L352) returns false and falls back to `sbgemm_` .   

>  do you still need the related modification in this PR? The modification in this PR is more efficient.

Agreed !

**Review by aditew01 — APPROVED** [review]

**Comment by Valentine233** [comment]
> update the test case. This is the slow fallback path, can we make sure it is tested in torch CI?

The fallback path can be tested with `ciflow/linux-aarch64`, which is already triggered.

**Review by mingfeima — APPROVED** [review]

**Comment by mingfeima** [comment]
are the CI failures real @Valentine233 ?

**Comment by Valentine233** [comment]
> are the CI failures real @Valentine233 ?

They are all unrelated. But I will rebase and re-run the CIs anyway.

File: aten/src/ATen/native/CPUBlas.cpp [hunk 1]
```diff
@@ -358,18 +358,25 @@ void gemm(
       int m_ = m, n_ = n, k_ = k, lda_ = lda, ldb_ = ldb, ldc_ = ldc;
       char transa_ = to_blas(transa), transb_ = to_blas(transb);
       float alpha_ = alpha, beta_ = beta;
-      int c_size = n_ * ldc_;
+      int c_size = n_ * m_;
       // C matrix in OpenBLAS sbgemm are of type "float" so we have to convert, copy and copy back.
-      std::vector<float> float_v(c, c + c_size);
+      std::vector<float> float_v(c_size, 0.0f);
```
Comment (ezyang) [inline aten/src/ATen/native/CPUBlas.cpp:363]: Thanks for zero initializing

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/native/CPUBlas.cpp [hunk 2]
```diff
@@ -358,18 +358,25 @@ void gemm(
...
               b, &ldb_,
               &beta_,
-              float_v.data(), &ldc_);
-      for (auto cv: float_v) {
-        *(c++) = c10::convert<at::BFloat16>(cv);
+              float_v.data(), &m_);
+      for (const auto j : c10::irange(n)) {
+        for (const auto i : c10::irange(m)) {
+          c[j * ldc_ + i] = c10::convert<at::BFloat16>(float_v[j * m_ + i]);
+        }
```
Comment (ezyang) [inline aten/src/ATen/native/CPUBlas.cpp:379]: o3 thinks that this is not semantics preserving when ldc != m. Specifically:

> The new code changes c_size from n * ldc_ to n * m_, copies into a float_v laid out with a stride of m_, and then passes &m_ as LDC to sbgemm_.  This silently assumes ldc == m (i.e. the C matrix is tightly packed).  If callers supply a larger leading dimension (common for alignment), the old path gave a correct answer; the new path (a) copies/updates only m_ elements per column, and (b) tells OpenBLAS that the destination is tightly packed.  For non-contiguous C this will yield memory corruption or wrong answers.

> Suggested fix: keep the tighter buffer only when ldc_ == m_ (fast path) and fall back to the original stride-aware copy otherwise; or allocate n_ * ldc_ and keep the original layout.  Either way, pass the correct LDC to sbgemm_.

> ... [To see this occur,] Run gemm with ldc > m (e.g. 8-wide padding) and enable ASan/UBSan.

> ...numerically you’ve discarded the caller’s padding.  That’s fine if the padding was unused, but it breaks equivalence with the reference path: ­a future optimisation that relies on writing into the padded region (e.g. fuse bias or activation to the right of m) will behave differently in the BF16 branch.

TBH, I'm not sure, I don't remember if the original padding here matters or not. Do you know?

Comment (Valentine233) [reply aten/src/ATen/native/CPUBlas.cpp:379]: Thanks! I suppose the function does not care about the paddings and we won't change anything else for `c`, except `n*m` with stride `ldc`.
